src/fonduer/supervision/labeler.py

```diff
@@ -356,30 +361,38 @@ def _after_apply(
 
     def get_gold_labels(
         self, cand_lists: List[List[Candidate]], annotator: Optional[str] = None
-    ) -> List[csr_matrix]:
-        """Load sparse matrix of GoldLabels for each candidate_class.
+    ) -> List[np.ndarray]:
+        """Load dense matrix of GoldLabels for each candidate_class.
 
         :param cand_lists: The candidates to get gold labels for.
         :type cand_lists: List of list of candidates.
         :param annotator: A specific annotator key to get labels for. Default
             None.
         :type annotator: str
-        :return: A list of MxN sparse matrix where M are the candidates and N is the
+        :return: A list of MxN dense matrix where M are the candidates and N is the
             annotators. If annotator is provided, return a list of Mx1 matrix.
-        :rtype: list[csr_matrix]
+        :rtype: list[np.ndarray]
         """
-        return get_sparse_matrix(self.session, GoldLabelKey, cand_lists, key=annotator)
+        return [
+            m.toarray() - 1  # convert to {-1, 0, ..., k-1}
+            for m in get_sparse_matrix(
+                self.session, GoldLabelKey, cand_lists, key=annotator
+            )
+        ]
 
-    def get_label_matrices(self, cand_lists: List[List[Candidate]]) -> List[csr_matrix]:
-        """Load sparse matrix of Labels for each candidate_class.
+    def get_label_matrices(self, cand_lists: List[List[Candidate]]) -> List[np.ndarray]:
```

Comment:
Can we add an option to support returning sparse/dense matrix here?

Reply:
Which one do you mean?
- sparse (ABSTAIN as -1)/dense (ABSTAIN as -1)
- sparse (ABSTAIN as 0)/dense (ABSTAIN as -1)

Reply:
The second one. We can have a wrapper and a unwrapper function for this.

Reply:
Please help me find a use case for such an option.
Once Fonduer shifts ABSTAIN to -1, users are forced to shift their labels too.
To avoid a confusion, I think anything that assumes ABSTAIN as 0 should not be exposed to users.

Reply:
Agree! I think it’s more clear to have a wrapper function (convert user behavior (ABSTAIN = -1) to fonduer internal behavior (ABSTAIN = 0)) and unwrapper function to do this then just hard code it. And I think this wrapper should not expose to user.

Reply:
Thanks. Will do.

Reply:
Resolved by 7fc08d985a65e08ff117d4bc66d372fbd30546e9.